File: backend/schemas/assignment.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from models.lecturer import AvailabilityDay, AvailabilitySlot
from models.session import SessionType


class AssignmentItem(BaseModel):
    session_id: str
    day: AvailabilityDay
    start_slot: AvailabilitySlot
    room_id: str


class AssignmentSaveRequest(BaseModel):
    assignments: list[AssignmentItem]


class AssignmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    assignment_id: str
    session_id: str
    unit_id: str
    unit_code: str
    unit_name: str
    session_type: SessionType
    duration: int
    lecturer_id: str | None
    lecturer_display_name: str
    # Unit 60: derived from the hidden session-student allocation rows.
    student_count: int
    # Internal validation payload only; the UI must not display tutorial
    # allocation membership.
    allocated_student_ids: list[str] = []
    day: AvailabilityDay
    start_slot: AvailabilitySlot
    room_id: str
    created_at: datetime
    updated_at: datetime

File: frontend/src/features/timetable/assignment.ts
import type { Day } from './slots'
import type { SessionType } from '@/lib/api/sessions'

export type SlotId = 's1' | 's2' | 's3' | 's4' | 's5' | 's6' | 's7'

// Frontend rendering model for a scheduled session.
// Non-overlapping placement is assumed — each (day, room, start_slot) is unique.
// Aligns with the future backend assignment DTO (Unit 32).
export interface TimetableAssignment {
  assignment_id?: string
  session_id: string
  unit_id: string
  unit_code: string
  unit_name: string
  session_type: SessionType
  duration: number
  lecturer_display_name: string
  // Optional: present when sourced from a schedulable session; saved
  // assignment rows do not carry it. Rendering does not depend on it yet
  // (Unit 67 wires session-level validation).
  lecturer_id?: string
  student_count: number
  // Optional internal validation payload for later units. Never displayed.
  allocated_student_ids?: string[]
  day: Day
  start_slot: SlotId
  room_id: string
}
